src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ReflectionReader.cs

```diff
@@ -0,0 +1,628 @@
+        {
+            object obj;
+            if (!classContract.CreateNewInstanceViaDefaultConstructor(out obj))
+            {
+                Type classType = classContract.UnderlyingType;
+                obj = XmlFormatReaderGenerator.UnsafeGetUninitializedObject(classType);
+            }
+
+            return obj;
+        }
+
+        private static object ResolveAdapterObject(object obj, ClassDataContract classContract)
+        {
+            Type objType = obj.GetType();
+            if (objType == Globals.TypeOfDateTimeOffsetAdapter)
+            {
+                obj = DateTimeOffsetAdapter.GetDateTimeOffset((DateTimeOffsetAdapter)obj);
+            }
+            else if (obj is IKeyValuePairAdapter)
+            {
+                obj = classContract.GetKeyValuePairMethodInfo.Invoke(obj, Array.Empty<object>());
+            }
+
+            return obj;
+        }
+
+        private object ReflectionReadCollectionCore(XmlReaderDelegator xmlReader, XmlObjectSerializerReadContext context, XmlDictionaryString collectionItemName, XmlDictionaryString collectionItemNamespace, CollectionDataContract collectionContract)
+        {
+            bool isArray = (collectionContract.Kind == CollectionKind.Array);
+
+            int arraySize = context.GetArraySize();
+            string objectId = context.GetObjectId();
+            object resultArray = null;
+            if (isArray && ReflectionTryReadPrimitiveArray(xmlReader, context, collectionItemName, collectionItemNamespace, collectionContract.UnderlyingType, collectionContract.ItemType, arraySize, out resultArray))
+            {
+                return resultArray;
+            }
+
+            if (arraySize != -1)
+            {
+                throw new NotImplementedException();
+            }
+
+            string itemName = GetCollectionContractItemName(collectionContract);
+            string itemNs = GetCollectionContractNamespace(collectionContract);
+            object resultCollection = ReflectionCreateCollection(collectionContract);
+            context.AddNewObject(resultCollection);
+            context.IncrementItemCount(1);
+
+            if (!ReflectionReadSpecialCollection(xmlReader, context, collectionContract, resultCollection))
+            {
+                ReflectionReadValueDelegate reflectionReadValueDelegate = null;
+                CollectionSetItemDelegate collectionSetItemDelegate = null;
+
+                int index = 0;
+                for (; index < int.MaxValue; index++)
+                {
+                    if (xmlReader.IsStartElement(collectionItemName, collectionItemNamespace))
+                    {
+                        context.IncrementItemCount(1);
+                        object collectionItem = ReflectionReadCollectionItem(xmlReader, context, collectionContract, collectionContract.ItemType, itemName, itemNs, ref reflectionReadValueDelegate);
+                        if (collectionSetItemDelegate == null)
```

Comment:
Although we're not guaranteed to hit this line (unlike in the earlier method), doing less work on each iteration when there's many data items is a better trade off than avoiding a little bit of work when there's no items. So unconditionally create delegate outside of loop.